File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.*;
import java.util.*;
import java.util.stream.*;

// 3 4

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        List<Integer> arr = new ArrayList<>();
        StringBuilder sb = new StringBuilder();
        Deque<Integer> stack = new ArrayDeque<>();
        int lastNum = 0;
        for (int i = 0; i < n; i++) {
            int cur = Integer.parseInt(br.readLine());
            while (stack.isEmpty() || stack.peek() < cur) {
                stack.push(++lastNum);
                sb.append("+\n");
            }

            if (stack.peek() != cur) {
                System.out.print("NO");
                System.exit(0);
            }

            stack.pop();
            sb.append("-\n");
        }
        System.out.print(sb);
    }
}

File: 백준/Silver/1874. 스택 수열/스택 수열.kt
import java.io.*
import java.util.*
import kotlin.math.*

// 문제: pop 결과로 부터 오름차순 push/pop 과정을 유추
/**
 * arr[i]를 pop하기 위해선, 
 * - arr[i]가 stack의 top에 존재해야 한다.
 * - 즉, stack의 top에 arr[i]가 없으면 해당 arr[i]까지 push 해야한다.
 *     - 근데, arr[i]가 top에 있는 것 보다 작으면? 불가능하다.
 */

fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()

    val sb = StringBuilder()
    val stack: Deque<Int> = ArrayDeque()
    var cur = 1
    repeat(n) {
        val x = readLine().toInt()

        while (stack.isEmpty() || stack.peek() < x) {
            stack.push(cur++)
            sb.append("+\n")
        }

        if (stack.peek() != x) {
            print("NO")
            System.exit(0)
        }

        stack.pop()
        sb.append("-\n")
    }

    print(sb)
}